hw/mcu/microchip/pic32mz2048efg100/src/hal_timer.c

```diff
@@ -0,0 +1,548 @@
+            break;
+        case 5:
+            IPC8CLR = _IPC8_T7IP_MASK | _IPC8_T7IS_MASK;
+            IPC8SET = 3 << _IPC8_T7IP_POSITION;
+            IFS1CLR = _IFS1_T7IF_MASK;
+            IEC1SET = _IEC1_T7IE_MASK;
+            break;
+        case 6:
+            IPC9CLR = _IPC9_T8IP_MASK | _IPC9_T8IS_MASK;
+            IPC9SET = 3 << _IPC9_T8IP_POSITION;
+            IFS1CLR = _IFS1_T8IF_MASK;
+            IEC1SET = _IEC1_T8IE_MASK;
+            break;
+        case 7:
+            IPC10CLR = _IPC10_T9IP_MASK | _IPC10_T9IS_MASK;
+            IPC10SET = 3 << _IPC10_T9IP_POSITION;
+            IFS1CLR = _IFS1_T9IF_MASK;
+            IEC1SET = _IEC1_T9IE_MASK;
+            break;
+    }
+}
+
+static void
+hal_timer_disable_int(int timer_num)
+{
+    switch (timer_num) {
+        case 0:
+            IEC0CLR = _IEC0_T2IE_MASK;
+            IFS0CLR = _IFS0_T2IF_MASK;
+            break;
+        case 1:
+            IEC0CLR = _IEC0_T3IE_MASK;
+            IFS0CLR = _IFS0_T3IF_MASK;
+            break;
+        case 2:
+            IEC0CLR = _IEC0_T4IE_MASK;
+            IFS0CLR = _IFS0_T4IF_MASK;
+            break;
+        case 3:
+            IEC0CLR = _IEC0_T5IE_MASK;
+            IFS0CLR = _IFS0_T5IF_MASK;
+            break;
+        case 4:
+            IEC0CLR = _IEC0_T6IE_MASK;
+            IFS0CLR = _IFS0_T6IF_MASK;
+            break;
+        case 5:
+            IEC1CLR = _IEC1_T7IE_MASK;
+            IFS1CLR = _IFS1_T7IF_MASK;
+            break;
+        case 6:
+            IEC1CLR = _IEC1_T8IE_MASK;
+            IFS1CLR = _IFS1_T8IF_MASK;
+            break;
+        case 7:
+            IEC1CLR = _IEC1_T9IE_MASK;
+            IFS1CLR = _IFS1_T9IF_MASK;
+            break;
+    }
+
+}
+
+static void
+compute_next_expiry(struct pic32_timer *bsp_timer)
+{
+    struct hal_timer *first = TAILQ_FIRST(&bsp_timer->hal_timer_queue);
+    if (first != NULL) {
+        bsp_timer->expiry = first->expiry;
+        bsp_timer->expiry *= bsp_timer->interrupt_per_tick;
+    }
+}
+
+static void
+handle_timer_list(int timer_num)
+{
+    int i;
+    int timers_to_remove = 0;
+    uint32_t current_tick = hal_timer_read(timer_num);
+    struct hal_timer *entry;
+
+    TAILQ_FOREACH(entry, &timers[timer_num].hal_timer_queue, link) {
+        if (entry->expiry <= current_tick) {
+            entry->cb_func(entry->cb_arg);
+            ++timers_to_remove;
+        } else {
+            break;
+        }
+    }
+
+    for (i = 0; i < timers_to_remove; ++i) {
+        TAILQ_REMOVE(&timers[timer_num].hal_timer_queue,
+                     TAILQ_FIRST(&timers[timer_num].hal_timer_queue), link);
+    }
+    if (timers_to_remove != 0) {
+        compute_next_expiry(&timers[timer_num]);
+    }
+}
+
+void
+__attribute__((interrupt(IPL3AUTO), vector(_TIMER_2_VECTOR)))
+timer2_isr(void)
+{
+    ++timers[0].interrupt_count;
+
+    if (!(TAILQ_EMPTY(&timers[0].hal_timer_queue)) &&
+        timers[0].expiry <= timers[0].interrupt_count) {
```

Comment:
I tested my code by setting a timer at 500kHz and I was already hitting some issues because the ```hal_timer_read``` took too much time to execute in the interrupt handler, that is why I used interrupt count to make ```hal_timer_read``` as fast as possible.

> Is there a register you can read for the time instead of relying on interrupt count?

Yes I can use ```PRx``` and ```TMRx```. 

I was looking at the nrf51xxx hal_timer implementation and I found out two things:
- The tick does not seem to be related to ```OS_TICKS_PER_SEC```. Having a 64-bit counter that increases at a fixed rate would be the current tick of the timer.
- The ```hal_config``` only seems to configure prescaler as the compare register is changed when a soft timer is added/removed in the list to generate as few interrupts as possible. From my understanding, the parameter ```freq_hz``` in ```hal_config``` does not specify the frequency of  interrupts generated by the timer module, but instead specifies the speed of increase of the ```TMRx``` register.

Are these two assumptions correct ?